File: src/main/java/com/fouan/actors/zombies/Zombie.java
package com.fouan.actors.zombies;

import com.fouan.events.SurvivorLostLifePoints;
import com.fouan.events.ZombieMoved;
import com.fouan.actors.Actor;
import com.fouan.actors.ActorId;
import com.fouan.game.view.GameView;
import com.fouan.zones.Zone;
import lombok.Getter;

@Getter
public abstract class Zombie extends Actor {

    protected final String name;
    protected final int damage;
    protected final int experienceProvided;

    protected Zombie(ActorId id, String name, int damage, int experienceProvided) {
        super(id);
        this.name = name;
        this.damage = damage;
        this.experienceProvided = experienceProvided;
    }

    @Override
    public void moveTo(Zone zoneToMove, GameView gameView) {
        gameView.fireEvent(
            new ZombieMoved(gameView.getCurrentTurn(), id, zoneToMove)
        );
    }

    @Override
    public void attack(ActorId actorId, GameView gameView) {
        gameView.fireEvent(
            new SurvivorLostLifePoints(gameView.getCurrentTurn(), actorId, id, damage)
        );
    }
}

File: src/main/java/com/fouan/actors/zombies/Zombie.kt
package com.fouan.actors.zombies;

import com.fouan.actors.Actor;
import com.fouan.actors.ActorId;

public abstract class Zombie extends Actor {

    protected final String name;
    protected final int damage;
    protected final int minDamageToDestroy;
    protected final int experienceProvided;

    protected Zombie(ActorId id, String name, int damage, int minDamageToDestroy, int experienceProvided, int actionsCount) {
        super(id, actionsCount);
        this.name = name;
        this.damage = damage;
        this.minDamageToDestroy = minDamageToDestroy;
        this.experienceProvided = experienceProvided;
    }

    public String getName() {
        return this.name;
    }

    public int getDamage() {
        return this.damage;
    }

    public int getMinDamageToDestroy() {
        return this.minDamageToDestroy;
    }

    public int getExperienceProvided() {
        return this.experienceProvided;
    }

    public String toString() {
        return "Zombie(name=" + this.getName() + ", damage=" + this.getDamage() + ", minDamageToDestroy=" + this.getMinDamageToDestroy() + ", experienceProvided=" + this.getExperienceProvided() + ")";
    }
}
